src/java.base/share/classes/java/util/Collections.java

```diff
@@ -2228,6 +2323,97 @@ public boolean equals(Object o) {
         public int hashCode() {
             synchronized (mutex) {return c.hashCode();}
         }
+
+        @java.io.Serial
+        private Object readResolve() {
+            if (c instanceof RegularEnumSetCompatible<?> es) {
+                return new SynchronizedRegularEnumSet<>(es);
+            }
+            if (c instanceof JumboEnumSetCompatible<?> es) {
+                return new SynchronizedJumboEnumSet<>(es);
+            }
+            return this;
+        }
+    }
+
+    /**
+     * @serial include
+     */
+    static final class SynchronizedRegularEnumSet<E extends Enum<E>>
+          extends SynchronizedSet<E>
+          implements RegularEnumSetCompatible<E> {
+        @java.io.Serial
+        private static final long serialVersionUID = -5185717517664879222L;
+
+        final RegularEnumSetCompatible<E> es;
+
+        SynchronizedRegularEnumSet(RegularEnumSetCompatible<E> es) {
+            super(es);
+            this.es = es;
+        }
+
+        SynchronizedRegularEnumSet(RegularEnumSetCompatible<E> es, Object mutex) {
+            super(es, mutex);
+            this.es = es;
+        }
+
+        @Override
+        public Class<E> elementType() {
+            // No need to be synchronized since elementType is never modified.
+            return es.elementType();
+        }
+
+        @Override
+        public long elements() {
+            synchronized (mutex) {
+                return es.elements();
+            }
+        }
+
+        @java.io.Serial
+        private Object writeReplace() {
+            return new SynchronizedSet<E>(es);
+        }
+    }
+
+    /**
+     * @serial include
+     */
+    static final class SynchronizedJumboEnumSet<E extends Enum<E>>
+          extends SynchronizedSet<E>
+          implements JumboEnumSetCompatible<E> {
+        @java.io.Serial
+        private static final long serialVersionUID = -2197803204861808334L;
+
+        final JumboEnumSetCompatible<E> es;
+
+        SynchronizedJumboEnumSet(JumboEnumSetCompatible<E> es) {
+            super(es);
+            this.es = es;
+        }
+
+        SynchronizedJumboEnumSet(JumboEnumSetCompatible<E> es, Object mutex) {
+            super(es, mutex);
+            this.es = es;
+        }
+
+        @Override
+        public Class<E> elementType() {
+            // No need to be synchronized since elementType is never modified.
+            return es.elementType();
+        }
+
+        @Override
+        public long[] elements() {
+            synchronized (mutex) {
+                return es.elements();
```

Comment:
Is this really safe? `elements()` in `JumboEnumSet` returns the backing array, so this returned array is mutable. Should perform `es.elements().clone()` for more safety.